marlowe-dashboard-client/src/Marlowe/Execution/Types.purs

```diff
@@ -1,60 +1,80 @@
 module Marlowe.Execution.Types
   ( ExecutionState
-  , PreviousState
+  , PreviousExecutionState
+  , CurrentExecutionState
   , PendingTimeouts
   , NamedAction(..)
   ) where
 
 import Prelude
 import Data.BigInteger (BigInteger)
+import Data.List (List)
 import Data.Map (Map)
 import Data.Maybe (Maybe)
 import Marlowe.Semantics (AccountId, Bound, ChoiceId, ChosenNum, Contract, Observation, Party, Payment, Slot, State, Token, TransactionInput, ValueId)
 
 type ExecutionState
-  = { previous :: Array PreviousState
-    , current ::
-        { state :: State
-        , contract :: Contract
-        }
+  = { previousExecutionStates :: Array PreviousExecutionState
+    , currentExecutionState :: CurrentExecutionState
     , mPendingTimeouts :: Maybe PendingTimeouts
     , mNextTimeout :: Maybe Slot
     }
 
--- This represents a previous step in the execution. The state property corresponds to the state before the
--- txInput was applied and it's saved as an early optimization to calculate the balances at each step.
-type PreviousState
-  = { txInput :: TransactionInput
+-- The current execution state is either the initial state of the contract, or a function of the
+-- last element in the `previous` array. It is saved as an early optimisation so that we don't need
+-- to recalculate payments and account balances all the the time.
```

Comment:
Not sure I understand this comment, and there is no `Either` involved. The execution state is both the current place of the execution and the previous states, at most the current step card is calculated from the current execution state or pending inputs, but that data lives in a different file, so I don't think we should put that type of comment here

Reply:
The comment needs rewriting then. The basic point I was wanting to make is that we only need to store the initial contract and state, and the array of transaction inputs - everything else is derived data that we store just so we don't have to recalculate it all the time. Specifically: the current `State` and `Contract` is a function of the previous `State` and `Contract` plus the previous `TransactionInput` (unless there is no previous `State`/`Contract`/`TransactionInput`, in which case it's the initial `State` and `Contract`).

Similarly with the comment below: the `Contract`, `State`, and `Payments` of `previous[n]` are a function of the `Contract`, `State`, and `TransactionInput` of `previous[n-1]` - unless `n == 0`, in which case they're the initial `State` and `Contract` (and the `Payments` are just `[]`).

Does that make any more sense?

Reply:
Well, I think a simpler explanation in the fields would do the trick and might be less confusing. When it applies we could just say

```purescript
type CurrentExecutionState
  = { contract :: Contract
    , state :: State
    -- The payments are store for performance issues, this could be derived from the output of computeTransaction
    , payments :: List Payment
    }
```